model/DataObject.php

```diff
@@ -1876,168 +1931,241 @@ public function has_many($component = null, $classOnly = true) {
 	}
 
 	/**
+	 * Return data for a specific has_many component.
+	 * @param string $component
+	 * @param bool $classOnly If this is TRUE, than any has_many relationships in the form "ClassName.Field" will have
+	 *        the field data stripped off. It defaults to TRUE.
+	 * @return string|false
+	 */
+	public function hasManyComponent($component, $classOnly = true) {
+		$hasMany = (array)Config::inst()->get($this->class, 'has_many', Config::INHERITED);
+
+		if($hasMany && array_key_exists($component, $hasMany)) {
+			$hasMany = $hasMany[$component];
+		} else {
+			return false;
+		}
+
+		return ($classOnly) ? preg_replace('/(.+)?\..+/', '$1', $hasMany) : $hasMany;
+	}
+
+	/**
+	 * @deprecated 4.0 Method has been renamed to manyManyExtraFields()
+	 * @param string $component
+	 * @return array
+	 */
+	public function many_many_extraFields($component = null) {
+		Deprecation::notice('4.0', 'Please use manyManyExtraFields() instead');
+		return $this->manyManyExtraFields($component);
+	}
+
+	/**
 	 * Return the many-to-many extra fields specification.
 	 *
 	 * If you don't specify a component name, it returns all
 	 * extra fields for all components available.
 	 *
 	 * @param string $component Name of component
-	 * @return array
+	 * @return array|null
 	 */
-	public function many_many_extraFields($component = null) {
-		$classes = ClassInfo::ancestry($this);
-
-		foreach($classes as $class) {
-			if(in_array($class, array('ViewableData', 'Object', 'DataObject'))) continue;
-			$relationName = null;
-
-			// Find extra fields for one component
-			if($component) {
-				$SNG_class = singleton($class);
-				$extraFields = $SNG_class->stat('many_many_extraFields');
-
-				// Extra fields are immediately available on this class
-				if(isset($extraFields[$component])) {
-					return $extraFields[$component];
-				}
+	public function manyManyExtraFields($component = null) {
+		if($component) {
+			return $this->manyManyExtraFieldsForComponent($component);
+		}
 
-				$manyMany = $SNG_class->stat('many_many');
-				$candidate = (isset($manyMany[$component])) ? $manyMany[$component] : null;
-				if($candidate) {
-					$SNG_candidate = singleton($candidate);
-					$candidateManyMany = $SNG_candidate->stat('belongs_many_many');
+		$items = (array)Config::inst()->get($this->class, 'many_many_extraFields', Config::INHERITED);
+		foreach($items as $k => $v) {
+			if(!is_array($v)) {
+				throw new LogicException("$this->class::\$many_many_extraFields has a bad entry: "
+					. var_export($k, true) . " => " . var_export($v, true)
+					. ". Each many_many_extraFields entry should map to a field specification array.");
+			}
+		}
 
-					// Find the relation given the class
-					if($candidateManyMany) foreach($candidateManyMany as $relation => $relatedClass) {
-						if($relatedClass == $class) {
-							$relationName = $relation;
-							break;
-						}
-					}
+		return count($items) ? $items : null;
```

Comment:
I was reluctant to change any return types in a minor release, though - could cause issues for anyone doing strict comparisons. Will change to `!empty()`